### PR TITLE
IOS XR cisco_bgp_neighbor_af support

### PR DESCRIPTION
**Pull request by HelenHZhang**
all beaker tests passed for xr, beaker tests for nexus passed before, is currently blocked by other issues,

### CODE REVIEW

**Comment by glennmatthews** [comment]
Looks like Travis CI is failing due to a number of Rubocop failures in the test code. Can you test (with rubocop 0.35.1) and fix any issues?

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb [hunk 1]
```diff
@@ -529,7 +592,9 @@ def build_manifest_bgp_nbr_af(tests, id)
                tests[id][:title_pattern])
   tests[id][:manifest] = "cat <<EOF >#{PUPPETMASTER_MANIFESTPATH}
   node 'default' {
-    cisco_bgp_neighbor_af { '#{tests[id][:title_pattern]}':
+    #{extra_config}
+
+    cisco_bgp_neighbor_af { '#{af[:asn]} #{af[:vrf]} #{af[:neighbor]} #{af[:afi]} #{af[:safi]}':
```
Comment (glennmatthews) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: Why the change away from using `:title_pattern` here?

**Comment by HelenHZhang** [comment]
I did use rubocop for the 3 files I have changed, none of violation is from my changes.
I was not sure I should fix it.
I will fix them, commit soon.

Helen

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb [hunk 2]
```diff
@@ -539,154 +604,170 @@ def build_manifest_bgp_nbr_af(tests, id)
...
-  bgp_nbr_remote_as(agent, tests[id][:remote_as]) if tests[id][:remote_as]
-
   tests[id][:ensure] = :present if tests[id][:ensure].nil?
-  tests[id][:show_pattern] = af_pattern(tests, id, af)
   tests[id][:resource_cmd] = puppet_resource_cmd(af)
 
-  # Build the manifest for this test
-  build_manifest_bgp_nbr_af(tests, id)
+  if platform == 'ios_xr' and af[:vrf] != 'default'
+    logger.info("\n--------\nSkip Case Address-Family ID: #{af} for ios_xr")
```
Comment (glennmatthews) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: This seems wrong - XR supports VRF neighbor_afs so why skip these tests?

**Comment by glennmatthews** [comment]
I'd expect some updates to the cisco_bgp_neighbor_af documentation in README.md - are none needed?

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb (hunk 1 above)
Comment (HelenHZhang) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: This is needed change due to "neighbor" parameter in Section 3. Test Pattern Testing.
@glennmatthews

Comment (glennmatthews) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: I don't understand. The whole point of section 3 is testing the specified title_pattern value - by ignoring it you are defeating the point of the test, are you not?

File: README.md [hunk 3]
```diff
@@ -590,10 +590,10 @@ BGP autonomous system number. Required. Valid values are String, Integer in ASPL
 VRF name. Required. Valid values are string. The name 'default' is a valid VRF representing the global bgp.
 
 ##### `afi`
-Address Family Identifier (AFI). Required. Valid values for Nexus are `ipv4` and `ipv6`.  Valid values for IOS XR are `ipv4`, `ipv6`, `vpnv4`, and `vpnv6`.
+Address Family Identifier (AFI). Required. Valid values for Nexus and IOS XR are `ipv4`, `ipv6`, `vpnp4`, `vpnp6` and `l2vpn.
```
Comment (glennmatthews) [inline README.md]: missing a trailing `` ` `` here (`` `l2vpn` ``)

File: README.md [hunk 4]
```diff
@@ -834,10 +834,10 @@ VRF name. Required. Valid values are string. The name 'default' is a valid VRF r
 Neighbor Identifier. Required. Valid values are string. Neighbors may use IPv4 or IPv6 notation, with or without a subnet mask.
 
 ##### `afi`
-Neighbor Address Family Identifier (AFI). Required. Valid values are string. Valid neighbor AFIs are `ipv4` and `ipv6`. Note that some AFI/SAFI address-families may not be supported with some neighbors; e.g. an ipv6 neighbor may not support an ipv4 multicast address-family.
+Neighbor Address Family Identifier (AFI). Required. Valid values are string. Valid neighbor AFIs are `ipv4`, `ipv6`, `vpnp4`, `vpnp6` and `l2vpn`. Note that some AFI/SAFI address-families may not be supported with some neighbors; e.g. an ipv6 neighbor may not support an ipv4 multicast address-family.
```
Comment (glennmatthews) [inline README.md]: "vpnv4", "vpnv6"

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb [hunk 5]
```diff
@@ -376,11 +394,11 @@
 
 tests['non_default_properties_ebgp_only'] = {
   :desc           => "2.9 Non Default Properties: 'ebgp' commands",
-  :title_pattern  => '2 yellow 3.3.3.3 ipv4 unicast',
-  :remote_as      => '2 yellow  3.3.3.3 3',
-  :manifest_props => "
-    as_override => true,
-  ",
+  :title_pattern  => "#{BgpLib::ASN} yellow 3.3.3.3 ipv4 unicast",
+  :remote_as      => 3,
```
Comment (chburket) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: What was the reasoning behind this change?  Is '2 yellow 3.3.3.3 3' a valid value for remote_as on nexus?

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb [hunk 6]
```diff
@@ -388,11 +406,11 @@
 
 tests['non_default_properties_ibgp_only'] = {
   :desc           => "2.10 Non Default Properties: 'ibgp' commands",
-  :title_pattern  => '2 green 2.2.2.2 ipv4 unicast',
-  :remote_as      => '2 green  2.2.2.2 2',
-  :manifest_props => "
-    route_reflector_client => true,
-  ",
+  :title_pattern  => "#{BgpLib::ASN} green 2.2.2.2 ipv4 unicast",
+  :remote_as      => "#{BgpLib::ASN}",
```
Comment (chburket) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: Same question, here.

**Comment by HelenHZhang** [comment]
fixed bgp.rb in cisco-network-node-utils, all beaker tests for ios-xr and nexus passed

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb [hunk 7]
```diff
@@ -539,154 +603,170 @@ def build_manifest_bgp_nbr_af(tests, id)
 
 # Wrapper for bgp_nbr_af specific settings prior to calling the
 # common test_harness.
-def test_harness_bgp_nbr_af(tests, id)
+def test_harness_bgp_nbr_af(tests, id, platform)
   af = af_title_pattern_munge(tests, id, 'bgp_neighbor_af')
   logger.info("\n--------\nTest Case Address-Family ID: #{af}")
 
-  # Set up remote-as if necessary
-  bgp_nbr_remote_as(agent, tests[id][:remote_as]) if tests[id][:remote_as]
```
Comment (glennmatthews) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: If the `bgp_nbr_remote_as` method is no longer used, please delete its definition.

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb [hunk 8]
```diff
@@ -492,36 +510,80 @@ def puppet_resource_cmd(af)
...
-       /address-family #{afi} #{safi}/]
+def get_dependency_manifest(platform, af, remote)
+  extra_config = ''
+  if platform == 'ios_xr'
+    # XR requires the following before a vrf AF can be configured:
+    #   1. a global router_id
+    #   2. a global address family
+    #   3. route_distinguisher configured on the vrf
+    #   4. remote-as is required for neightbor
+    remote = 2 if remote.nil?
```
Comment (chburket) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: Maybe better to add a default value in the function args?  
def get_dependency_manifest(platform, af, remote=2)

And can you rename ``remote`` to ``remote_as``?

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb [hunk 9]
```diff
@@ -67,12 +68,10 @@
 # Top-level keys set by caller:
 # tests[:master] - the master object
 # tests[:agent] - the agent object
-# tests[:show_cmd] - the common show command to use for test_show_run
 #
 tests = {
```
Comment (glennmatthews) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: This is gonna be *fun* to merge with the changes from #176 - but we'll burn that bridge when we come to it. :persevere:

File: README.md (hunk 3 above)
Comment (HelenHZhang) [inline README.md]: fixed, thanks for catching it.

File: README.md (hunk 4 above)
Comment (HelenHZhang) [inline README.md]: fixed, thanks for catching it.

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb (hunk 5 above)
Comment (HelenHZhang) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: In the old way (before rebase), remote_as has a full qualified string (including asn, vrf, neighbor), then calling "bgp_nbr_remote_as" using vsh command to configure the remote-as. We don't need it any more, we just need remote-as parameter

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb (hunk 7 above)
Comment (HelenHZhang) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: I deleted it before, it came back after rebase. I did it again.

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb (hunk 8 above)
Comment (HelenHZhang) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: "remote" is not necessary for nexus. I need to check the value whether it's nil for nexus.
I can rename it as "remote-as".

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb [hunk 10]
```diff
@@ -492,36 +510,80 @@ def puppet_resource_cmd(af)
...
+  if platform == 'ios_xr'
+    # XR requires the following before a vrf AF can be configured:
+    #   1. a global router_id
+    #   2. a global address family
+    #   3. route_distinguisher configured on the vrf
+    #   4. remote-as is required for neightbor
+    remote = 2 if remote.nil?
+    if af[:vrf] == 'default'
+      extra_config = "
+      cisco_bgp { '#{af[:asn]} #{af[:vrf]}':
```
Comment (chburket) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: If vrf is default, why do you need to configure a bgp instance as a dependency?  Shouldn't the default instance already be configured?

Comment (chburket) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: Looking at this some more, it seems like you are doing more "dependency setup" than is required.

In general, I don't think you have to create any of the "parent" instances (bgp, bgp_af) unless there are non-default values you need to set (like the bgp router_id when configuring an af under a vrf).

I don't know the details behind each of these, so I could certainly be wrong, but it seems like the following would be enough:

On XR:
- configure a bgp instance with a router_id (not sure if this is needed for both default & non-default vrf cases)
- configure a "parent" vpnv4 af
- configure any route-policies you need
- when vrf != default, configure a bgp vrf instance with a route_distinguisher set
- not sure about the neighbor with a remote_as, it looks like it is needed based on your comments/code

It looks like the neighbor is required for nexus, as well, so maybe that should just be moved out of the "if ios_xr" block?

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb [hunk 11]
```diff
@@ -539,154 +603,170 @@ def build_manifest_bgp_nbr_af(tests, id)
...
-  # Set up remote-as if necessary
-  bgp_nbr_remote_as(agent, tests[id][:remote_as]) if tests[id][:remote_as]
-
   tests[id][:ensure] = :present if tests[id][:ensure].nil?
-  tests[id][:show_pattern] = af_pattern(tests, id, af)
   tests[id][:resource_cmd] = puppet_resource_cmd(af)
 
-  # Build the manifest for this test
-  build_manifest_bgp_nbr_af(tests, id)
+  if platform == 'ios_xr' && af[:vrf] != 'default'
```
Comment (chburket) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: This looks like we aren't testing nbr af under non-default vrfs.  Is that intended?

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb [hunk 12]
```diff
@@ -539,154 +603,170 @@ def build_manifest_bgp_nbr_af(tests, id)
...
-  test_harness_bgp_nbr_af(tests, 'non_default_properties_A2')
-  test_harness_bgp_nbr_af(tests, 'non_default_properties_A3')
-  test_harness_bgp_nbr_af(tests, 'non_default_properties_D')
-  test_harness_bgp_nbr_af(tests, 'non_default_properties_M')
-  test_harness_bgp_nbr_af(tests, 'non_default_properties_N')
-  test_harness_bgp_nbr_af(tests, 'non_default_properties_S1')
-  test_harness_bgp_nbr_af(tests, 'non_default_properties_S2')
-  test_harness_bgp_nbr_af(tests, 'non_default_properties_S3')
-  test_harness_bgp_nbr_af(tests, 'non_default_properties_W')
+  init_bgp(master, agent)
```
Comment (chburket) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: If you're calling init_bgp here, no need to call cleanup_bgp right before it (init does a cleanup and an init)

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb (hunk 2 above)
Comment (HelenHZhang) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: On XR, I tried to configure manually. Two configuration works:

1)
router bgp 1
 address-family ipv4 unicast
 neighbor 1.1.1.1
 remote-as 2                  <<<< this is required
 address-family ipv4 unicast  <<<< parent-child address-family must match

2).
router bgp 1
 bop router-id 1.2.3.4
 address-family ipv4 unicast
 vrf blue
  rd auto
  neighbor 1.1.1.1
   remote-as 2               <<<< this is optional
                             <<<< can not configure address-family

For non-default VRF, address-family is not allowed.

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb (hunk 12 above)
Comment (HelenHZhang) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: will fix it.

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb (hunk 11 above)
Comment (HelenHZhang) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: yes, see other comment under this same code

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb (hunk 2 above)
Comment (glennmatthews) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: Works on my end:

```
router bgp 22
 default-information originate
 address-family ipv4 unicast
 !
 address-family vpnv4 unicast
 !
 af-group group1 address-family ipv4 unicast
 !
 neighbor 1.1.1.1
  remote-as 23
  capability additional-paths send
  address-family ipv4 unicast
  !
  address-family ipv4 labeled-unicast
  !
  address-family vpnv4 unicast
   import reoriginate stitching-rt
  !
 !
 vrf red
  rd 1:1
  address-family ipv4 unicast
  !
  neighbor 2.2.2.2
   remote-as 23
   address-family ipv4 unicast   <----------<<
    allowas-in 10
   !
  !
 !
!
```

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb [hunk 13]
```diff
@@ -492,36 +513,89 @@ def puppet_resource_cmd(af)
...
+        router_id                              => '1.2.3.4',
+      }
+      cisco_bgp_af { '#{af[:asn]} default #{global_afi} #{af[:safi]}':
+        ensure                                 => present,
+      }
+      cisco_bgp { '#{af[:asn]} #{af[:vrf]}':
+        ensure                                 => present,
+        route_distinguisher                    => auto,
+      }
+      cisco_bgp_af { '#{af[:asn]} #{af[:vrf]} #{af[:afi]} #{af[:safi]}':
```
Comment (glennmatthews) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: You have some duplication here between the vrf and non-vrf cases - additionally, the XR prerequisite configs should be harmless to add on Nexus so you probably don't need to have that duplication either. Something like:

```ruby
def get_dependency_manifest(platform, af, remote_as)
  extra_config = ''
  # XR requires the following before a vrf AF can be configured:
  #   1. a global router_id
  #   2. a global address family
  #   3. route_distinguisher configured on the vrf
  #   4. remote-as is required for neightbor
  # Nexus doesn't require these but it doesn't hurt to include them
  remote_as = 2 if remote_as.nil? && platform == ios_xr
  if af[:vrf] != 'default'
    if af[:afi] == 'ipv6'
      global_afi = 'vpnv6'
    else
      global_afi = 'vpnv4'
    end
    extra_config = "
      cisco_bgp { '#{af[:asn]}':
        ensure                                 => present,
        router_id                              => '1.2.3.4',
      }
      cisco_bgp_af { '#{af[:asn]} default #{global_afi} #{af[:safi]}':
        ensure                                 => present,
      }
    "
  end
  extra_config += "
      cisco_bgp { '#{af[:asn]} #{af[:vrf]}':
        ensure                                 => present,
        route_distinguisher                    => auto,
      }
      cisco_bgp_af { '#{af[:asn]} #{af[:vrf]} #{af[:afi]} #{af[:safi]}':
        ensure                                 => present,
      }
    "
  if remote_as
    extra_config += "
      cisco_bgp_neighbor { '#{af[:asn]} #{af[:vrf]} #{af[:neighbor]}':
        ensure                                 => present,
        remote_as                              => #{remote_as},
      }"
  end
  if platform == "ios_xr"
    extra_config += "
      cisco_command_config { 'policy_config':
        command => '
          route-policy rm_in
            end-policy
          route-policy rm_out
            end-policy'
      }"
  end
  extra_config
end
```

**Comment by chburket** [comment]
:+1: 

Might want to make note of the different types of changes you needed to make, so you can make them again when we merge in #176

File: tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb (hunk 13 above)
Comment (HelenHZhang) [inline tests/beaker_tests/cisco_bgp_neighbor_af/test_bgpneighboraf.rb]: revised based on your comment, plus moving "route_distinguisher" clause into non default vrf branch.
